src/lxml/sax.py

```diff
@@ -232,14 +235,25 @@ def _recursive_saxify(self, element, prefixes):
         if element.tail:
             content_handler.characters(element.tail)
 
-    def _build_qname(self, ns_uri, local_name, prefixes, new_prefixes):
+    def _build_qname(self, ns_uri, local_name, prefixes, new_prefixes,
+                     element_prefixes=None):
         if ns_uri is None:
             return local_name
+
         try:
             prefix = prefixes[ns_uri]
         except KeyError:
-            prefix = prefixes[ns_uri] = 'ns%02d' % len(prefixes)
+            # New prefix
+            if element_prefixes and ns_uri in element_prefixes:
+                prefix = element_prefixes[ns_uri]
+            else:
+                # Unknown prefix, generate one
```

Comment:
While writing tests I wasn't able to make a element tree that actually has an undefined namespace. I don't know if this code then is unnecessary, or if I'm not good enough at bending etree's to my will. :-)

Reply:
That code is unnecessary, I added a test and the ns0 is generated elsewhere.